OpenRA.Mods.Common/Widgets/Logic/Ingame/LeaveMapLogic.cs

```diff
@@ -33,6 +34,11 @@ public LeaveMapLogic(Widget widget, World world, OrderManager orderManager)
 		{
 			this.orderManager = orderManager;
 
+			if (world.LocalPlayer.WinState == WinState.Won)
+			{
+				CampaignProgress.SaveProgress(world.LocalPlayer.Country.Name);
```

Comment:
This will be called for skirmish and multiplayer games too, which doesn't seem optimal at all.

It would be much cleaner if you exposed a Lua API that the map script could interact with for setting and querying campaign status.